python/cudf/cudf/utils/utils.py

```diff
@@ -464,3 +466,64 @@ def search_range(start, stop, x, step=1, side="left"):
 
     length = (stop - start) // step
     return max(min(length, i), 0)
+
+
+# Utils for using appropriate dispatch for array functions
+def get_appropriate_dispatched_func(
+    cudf_submodule, cupy_submodule, func, args, kwargs
+):
+    fname = func.__name__
+    if hasattr(cudf_submodule, fname):
+        cudf_func = getattr(cudf_submodule, fname)
+        return cudf_func(*args, **kwargs)
+
+    elif hasattr(cupy_submodule, fname):
+        cupy_func = getattr(cupy_submodule, fname)
+        # Handle case if cupy impliments it as a numpy function
+        # Unsure if needed
+        if cupy_func is func:
+            return NotImplemented
+
+        cupy_compatible_args = get_cupy_compatible_args(args)
+        cupy_output = cupy_func(*cupy_compatible_args, **kwargs)
+        return cast_to_appropriate_cudf_type(cupy_output)
+    else:
+        return NotImplemented
+
+
+def cast_to_appropriate_cudf_type(val):
+    # TODO Handle scalar
+    if val.ndim == 0:
+        return cudf.Scalar(val).value
```

Comment:
Unsure what should I output when the output of the NumPy function is a scalar but dispatched cupy function creates a zero-dimensional array.

Reply:
I think returning `.value` seems to be a valid approach for interop with other libraries. @brandon-b-miller do you think we must return only a `cudf.Scalar` instead?

Reply:
calling `.value` here will give you back a host side numpy scalar. Is the intention to create a scalar from a 0-D cupy array without incurring a host transfer?

Reply:
The goal here was to emulate the `cuDF.Series` functions that return scalar and just follow that behavior. We currently return `numpy scalars` ( See the below example).   Happy to change it to what ever you guys recommend. 

```python
>>> import cudf
>>> o = cudf.Series([1,2,3]).sum()
>>> type(o)
<class 'numpy.int64'>
```

Reply:
Ah, I see. For at least one release most likely, we'll keep returning numpy scalars for reduction ops like `sum`. This code should work fine, under the hood constructing a `cudf.Scalar` will call `val.item()` in the 0-D cupy array case and return that on the host.

Reply:
Gotcha, Thanks a lot for the inputs @brandon-b-miller ,  @galipremsagar . I am leaving the code as-is. We can change this to return `cudf scalar` once we change behavior for all reduction ops. Resolving this.